### PR TITLE
feat: rust key derivation

### PR DESCRIPTION
**Pull request by bkioshn**
# Description

Thanks for contributing to the project!
Please fill out this template to help us review your changes.

## Related Issue(s)

List the issue numbers related to this pull request.

> e.g., Closes #1, Resolves #1  Fixes #1

## Description of Changes

Provide a clear and concise description of what the pull request changes.

## Breaking Changes

Describe any breaking changes and the impact.

## Screenshots

If applicable, add screenshots to help explain your changes.

## Related Pull Requests

If applicable, list any related pull requests.

> e.g., #1, #1

## Please confirm the following checks

* [ ] My code follows the style guidelines of this project
* [ ] I have performed a self-review of my code
* [ ] I have commented my code, particularly in hard-to-understand areas
* [ ] I have made corresponding changes to the documentation
* [ ] My changes generate no new warnings
* [ ] I have added tests that prove my fix is effective or that my feature works
* [ ] New and existing unit tests pass locally with my changes
* [ ] Any dependent changes have been merged and published in downstream module

### CODE REVIEW

File: catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+/// - `mnemonic`: A string representing the mnemonic.
+/// - `passphrase`: An optional string representing the passphrase.
+///
+/// # Returns
+///
+/// Returns the `XPrv` extended private key as a `Result`.
+/// If the conversion is successful, it returns `Ok` with the extended private key
+/// (`XPrv`).
+pub fn mnemonic_to_xprv(mnemonic: String, passphrase: Option<String>) -> anyhow::Result<XPrvBytes> {
+    let xprv = mnemonic_to_xprv_helper(mnemonic, passphrase)?;
```
Comment (dtscalac) [inline catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs:31]: Make sure to run it in a thread_pool [like here](https://github.com/input-output-hk/catalyst-voices/blob/main/catalyst_voices_packages/catalyst_key_derivation/rust/src/api/simple.rs#L12) to avoid blocking the main thread on Flutter side

File: catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+///
+/// - `xprivate_key_bytes`: An extended private key of type `XPrvBytes`.
+/// - `path`: Derivation path. eg. m/0/2'/3 where ' represents hardened derivation.
+///
+/// # Returns
+///
+/// Returns the `XPrv` extended private key as a `Result`.
+/// If the derivation path is successful, it returns `Ok` with the extended private key
+/// (`XPrv`).
+pub fn derive_xprivate_key(xprivate_key_bytes: XPrvBytes, path: String) -> anyhow::Result<XPrvBytes> {
```
Comment (dtscalac) [inline catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs:72]: Please run in a thread_pool

File: catalyst_voices_packages/catalyst_key_derivation/example/lib/main.dart
Comment (apskhem) [inline catalyst_voices_packages/catalyst_key_derivation/example/lib/main.dart:1]: I think we should have a test (that has expect results) on the Flutter side to show that, at least, this bridge is working properly.

Comment (bkioshn) [reply catalyst_voices_packages/catalyst_key_derivation/example/lib/main.dart:1]: Agree, will leave it to frontend to it

File: catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+) -> anyhow::Result<bool> {
+    let result = spawn_blocking_with(
+        move || -> anyhow::Result<bool> {
+            let verified_sig: Signature<SignatureBytes> = match Signature::from_slice(&signature) {
+                Ok(sig) => sig,
+                // Invalid signature, force return false.
+                Err(_) => return Ok(false),
+            };
+            let xprv = XPub::from_bytes(xpub_bytes);
+            Ok(check_signature_xpub_helper(&xprv, &data, verified_sig))
```
Comment (dtscalac) [inline catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs:280]: xpub?

```suggestion
            let xpub = XPub::from_bytes(xpub_bytes);
            Ok(check_signature_xpub_helper(&xpub, &data, verified_sig))
```

**Comment by dtscalac** [comment]
The dart part looks good to me. Would be nice if a rust engineer has a proper look at the rust code.

**Review by dtscalac — APPROVED** [review]

File: catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+/// Compose of:
+/// - 64 Bytes: extended Ed25519 secret key
+/// - 32 Bytes: chain code
+pub type XPrvBytes = [u8; 96];
+
+/// Extended public key bytes type.
+pub type XPubBytes = [u8; 64];
+
+/// Signature bytes type.
+pub type SignatureBytes = [u8; 64];
```
Comment (minikin) [inline catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs:26]: What do you think about using `newtypes` instead of `type aliases`? It could help prevent accidentally mixing up the different key types.

File: catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+
+/// Helper function for `derive_xprv`.
+fn derive_xprv_helper(xprv: XPrv, path: &str) -> anyhow::Result<XPrv> {
+    let Ok(derivation_path) = path.parse::<DerivationPath>() else {
+        return Err(anyhow::anyhow!("Invalid derivation path: {path}"));
+    };
+    let key = derivation_path.iter().fold(xprv, |xprv, child_num| {
+        if child_num.is_hardened() {
+            // i >= 2^31 is a hardened derivation
+            xprv.derive(DerivationScheme::V2, child_num.index() | 0x80_00_00_00)
```
Comment (minikin) [inline catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs:135]: Can you explain what `0x80_00_00_00` means, please?

File: catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+                Ok(xprv) => xprv,
+                Err(e) => return Err(anyhow::anyhow!(e)),
+            };
+            Ok(xpublic_key_helper(&xprv))
+        },
+        FLUTTER_RUST_BRIDGE_HANDLER.thread_pool(),
+    )
+    .await??;
+
+    Ok(Into::<[u8; 64]>::into(xpub))
```
Comment (minikin) [inline catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs:169]: Another reason why we should consider new types instead of type aliases...

**Review by minikin — CHANGES_REQUESTED** [review]
Overall looks great, I left a few suggestion.

**Comment by minikin** [comment]
@bkioshn To improve security, maybe we should also implement `auto-zeroizing for private keys on the drop.` What do you think?

**Comment by bkioshn** [comment]
> To improve security, maybe we should also implement `auto-zeroizing for private keys on the drop.` What do you think?

@minikin Yes, that would be great, but should it be in here or should it be handle in frontend?

File: catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs (hunk 5 above)
Comment (bkioshn) [reply catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs:135]: `0x80_00_00_00` is equivalent to 2^31 which indicate harden derivation
Maybe this is more clear https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki

**Comment by minikin** [comment]
> Yes, that would be great, but should it be in here or should it be handle in frontend?

@bkioshn I suggest we do it in Rust because we can explicitly control it. 
But it's a good catch, and we probably need to do something similar in Dart. 

PS: It's interesting to see how `auto-zeroing` on their rust side will work with a Dart.

File: catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs (hunk 5 above)
Comment (minikin) [reply catalyst_voices_packages/catalyst_key_derivation/rust/src/api/key_derivation/mod.rs:135]: Can we please consider adding a comment?
Something like this:

```txt
// Apply hardened derivation by setting the highest bit (0x80000000).
// This ensures enhanced security by using a private derivation scheme,
// preventing parent key leakage even if child keys are compromised.
// Format: m/i' where i >= 0x80000000 (hardened index)
```
It could be worth adding the URL that you posted, too.

**Review by stevenj — CHANGES_REQUESTED** [review]
As a general rule, we should NOT be embedding autogenerated files inside the git repo.
How are they generated?
Do i need to read a book about how to re-generate them if the bridge tools are updated?
Do i need to update the Rust code independently of the dart code, or is it generated together?
All autogenerated files should, ideally:
1. Be in a .gitignore file so they are never committed to the repo. (And the .gitignore should have a comment explaining this),
2. In CI, they are generated as part of the build, every time, with the version of the tools fixed in CI.
3. If they are needed locally, an Earthfile or Justfile can be used to generate local versions for testing (this needs to be documented).  Ideally this uses the same version of tools fixed in CI, so its better if its built with an Earthfile, but a Justfile could run earthly to do this.

As we add more rust bound libraries, the maintenance burden of ensuring these bindings are always up-to-date will become overwhelming, and slow down developement.

**Comment by Mr-Leshiy** [comment]
@bkioshn as we are not running any actual rust lints (`clippy`, `fmt` etc.) and doing any actual standard rust building procedure, I would remove from this PR such files like `.cargo/config.toml`, `deny.toml`, `clippy.toml`, `rustfmt.toml` and `.config/nextest.toml`.

**Comment by dtscalac** [comment]
> As a general rule, we should NOT be embedding autogenerated files inside the git repo.
> 
>     * How are they generated?
> 
>     * Do i need to read a book about how to re-generate them if the bridge tools are updated?
> 
>     * Do i need to update the Rust code independently of the dart code, or is it generated together?
> 
>     * What else would I need to know to maintain this?
> 
> 
> All autogenerated files should, ideally:
> 
>     1. Be in a .gitignore file so they are never committed to the repo. (And the .gitignore should have a comment explaining this),
> 
>     2. In CI, they are generated as part of the build, every time, with the version of the tools fixed in CI.
> 
>     3. If they are needed locally, an Earthfile or Justfile can be used to generate local versions for testing (this needs to be documented).  Ideally this uses the same version of tools fixed in CI, so its better if its built with an Earthfile, but a Justfile could run earthly to do this.  I should not need to install any tools to do this autogeneration outside of Earthly, or to develop with this.
> 
> 
> As we add more rust bound libraries, the maintenance burden of ensuring these bindings are always up-to-date will become overwhelming, and slow down development.

For apps and internal packages this is true however publicly exposed packages (i.e. via [pub.dev](pub.dev)) DO include the auto-generated files because developers who use them assume they will work out of the box without any extra steps. Otherwise the package would be unusable without forking it, generating these files and making the app talk to the forked version.

`catalyst_key_derivation` is going to be published on pub.dev because it's needed for `catalyst_cardano_serialization` which has been published already. The `catalyst_cardano_serialization` provides RBAC stuff and needs key derivation and signing which key_derivation package provides.

For example, the [json_serializable](https://pub.dev/packages/json_serializable) is one of the most popular code generators. Here are the packages that depend on the `json_serializable`: https://pub.dev/packages?q=dependency%3Ajson_serializable. All of these include the auto-generated files in git.

Screenshot 2024-11-05 at 12 48 29

TLDR: apps and internal packages may not include generated files but published packages must include them. Rust auto-generated files can be skipped but dart auto-generated files cannot be skipped.

**Comment by stevenj** [comment]
> TLDR: apps and internal packages may not include generated files but published packages must include them. Rust auto-generated files can be skipped but dart auto-generated files cannot be skipped.

Understood, but we need to clearly document how they need to be generated and updated.  And we should have a CI check to ensure they are not out of date.
They are probably not as bad as the API auto-generation because they will not be changing often.
Also, when your working on the API's you know your working on the bridge, its not a side effect.

File: catalyst_voices/packages/libs/catalyst_key_derivation/Earthfile [hunk 7]
```diff
@@ -0,0 +1,23 @@
+VERSION 0.8
+
+IMPORT github.com/input-output-hk/catalyst-ci/earthly/flutter_rust_bridge:v3.2.24 AS flutter_rust_bridge
+
+builder:
+    FROM flutter_rust_bridge+builder
+    COPY . .
+
+# Generated necessary files for running Flutter web locally and save it locally.
+code-generator:
```
Comment (Mr-Leshiy) [inline catalyst_voices/packages/libs/catalyst_key_derivation/Earthfile:10]: I would rather split this target into two `code-generator` and `code-generator-local` which would be more convenient to run just `code-generator-local` from command line.

Though I would make it like this
```Earthfile
code-generator-base:
    FROM +builder
    DO flutter_rust_bridge+CODE_GENERATOR_WEB
    
code-generator: 
    FROM +code-generator-base
    
   SAVE ARTIFACT ./assets/js assets_js
   SAVE ARTIFACT ./rust/src/frb_generated.rs frb_generated.rs
   SAVE ARTIFACT ./lib/src src

code-generator-local: 
    FROM +code-generator-base
    
    SAVE ARTIFACT ./assets/js AS LOCAL ./assets/js
    SAVE ARTIFACT ./rust/src/frb_generated.rs AS LOCAL ./rust/src/frb_generated.rs
    SAVE ARTIFACT ./lib/src AS LOCAL ./lib/src
```

File: catalyst_voices/packages/libs/catalyst_key_derivation/rust/src/lib.rs [hunk 8]
```diff
@@ -1,2 +1,20 @@
+//! Catalyst Key Derivation
+
+// cspell: words uninlined
+
 pub mod api;
+#[allow(
```
Comment (Mr-Leshiy) [inline catalyst_voices/packages/libs/catalyst_key_derivation/rust/src/lib.rs:6]: Just a question, this disabled lints for the autogenerated code right ?

Comment (bkioshn) [reply catalyst_voices/packages/libs/catalyst_key_derivation/rust/src/lib.rs:6]: Yep, without this it won't pass the CI.

**Review by Mr-Leshiy — CHANGES_REQUESTED** [review]
Reminder to resolve this comment https://github.com/input-output-hk/catalyst-voices/pull/1063/files#r1836126375

**Review by Mr-Leshiy — APPROVED** [review]
